In [2]:
import pandas as pd

### MUESTRA DE DATOS

# 100K DATASET

In [3]:

#100K
links_100K = pd.read_csv('BBDD_100K/links.csv')
movies_100K = pd.read_csv('BBDD_100K/movies.csv')
ratings_100K = pd.read_csv('BBDD_100K/ratings.csv')
tags_100K = pd.read_csv('BBDD_100K/tags.csv')

In [4]:
print(links_100K.head())
'''
movieId: ID de la película (de MovieLens).
imdbId: ID de la película en IMDb.
tmdbId: ID de la película en TMDb.
'''

   movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0


'\nmovieId: ID de la película (de MovieLens).\nimdbId: ID de la película en IMDb.\ntmdbId: ID de la película en TMDb.\n'

In [5]:
print(movies_100K.head())
'''
movieId: ID de la película.
title: Título de la película, incluyendo el año.
genres: Lista de géneros separados por "|".
'''

   movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy  


'\nmovieId: ID de la película.\ntitle: Título de la película, incluyendo el año.\ngenres: Lista de géneros separados por "|".\n'

In [6]:
print(ratings_100K.head())
'''
userId: ID del usuario.
movieId: ID de la película.
rating: Calificación de la película en una escala de 0.5 a 5.0.
timestamp: Momento en que se hizo la calificación.
'''

   userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931


'\nuserId: ID del usuario.\nmovieId: ID de la película.\nrating: Calificación de la película en una escala de 0.5 a 5.0.\ntimestamp: Momento en que se hizo la calificación.\n'

In [7]:
print(tags_100K.head())
'''
userId: ID del usuario.
movieId: ID de la película.
tag: Etiqueta asignada.
timestamp: Momento en que se añadió la etiqueta.
'''

   userId  movieId              tag   timestamp
0       2    60756            funny  1445714994
1       2    60756  Highly quotable  1445714996
2       2    60756     will ferrell  1445714992
3       2    89774     Boxing story  1445715207
4       2    89774              MMA  1445715200


'\nuserId: ID del usuario.\nmovieId: ID de la película.\ntag: Etiqueta asignada.\ntimestamp: Momento en que se añadió la etiqueta.\n'

### DATOS CONCRETOS

In [8]:
df_100K = ratings_100K.copy()
df_100K.head()

,userId,movieId,rating,timestamp
0,1,1,4.0,964982703
1,1,3,4.0,964981247
2,1,6,4.0,964982224
3,1,47,5.0,964983815
4,1,50,5.0,964982931


Cuantos usuarios, peliculas y ratings tenemos?

Usuarios

In [9]:
df_100K['userId'].unique()

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

610 usuarios en total

In [10]:
len(df_100K['userId'].unique())

610

Cuantas películas tenemos?

In [11]:
df_100K['movieId'].unique()

array([     1,      3,      6, ..., 160836, 163937, 163981], dtype=int64)

9724 películas en total 

In [12]:
len(df_100K['movieId'].unique())

9724

In [13]:
df_100K['rating'].head()

0    4.0
1    4.0
2    4.0
3    5.0
4    5.0
Name: rating, dtype: float64

Las valoraciones van desde los 0.5 a 5

In [14]:
print("Min rating = "+str(df_100K['rating'].min()))
print("Max rating = "+str(df_100K['rating'].max()))

Min rating = 0.5
Max rating = 5.0


El usuario que mas ratings ha puesto ha valorado 2698 películas, el que menos 20

In [15]:
# Suponiendo que el DataFrame se llama 'ratings'
ratings_count_per_user = df_100K.groupby('userId')['rating'].count()

# Obtener el máximo y el mínimo de ratings por usuario
max_ratings = ratings_count_per_user.max()
min_ratings = ratings_count_per_user.min()

print(f"Máximo de ratings por usuario: {max_ratings}")
print(f"Mínimo de ratings por usuario: {min_ratings}")

Máximo de ratings por usuario: 2698
Mínimo de ratings por usuario: 20


100836 valoraciones distintas

In [18]:
len(df_100K['rating'])

100836

Tenemos valores nulos?

In [19]:
df_100K['userId'].isnull().sum()



0

In [20]:
df_100K['rating'].isnull().sum()

0

In [21]:
df_100K['movieId'].isnull().sum()

0

No hay valores nulos

### Pequeño resumen (BBDD 100K)

Tenemos una base de datos con 610 usuarios y 9724 películas, donde cada usuario ha votado como minimo 20 películas y el que mas 2698, con un total de 100836 valoraciones, sin valores nulos y con ratings que van desde 0.5 a 5 de puntuación.

## VISUALIZACIÓN DE LA ESCASEZ DE DATOS

Para visualizar el problema de la escasez de datos voy a crear la matriz de ratings de usuarios y películas

In [22]:
indice=list(df_100K['userId'].unique())
columnas=list(df_100K['movieId'].unique())
indice=sorted(indice)
columnas=sorted(columnas)
 
escasez_100K=pd.pivot_table(data=df_100K,values='rating',index='userId',columns='movieId')

In [23]:
escasez_100K

movieId,1,2,3,4,5,6,7,8,9,10,...,193565,193567,193571,193573,193579,193581,193583,193585,193587,193609
userId,,,,,,,,,,,,,,,,,,,,,
1,4.0,NaN,4.0,NaN,NaN,4.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5,4.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
606,2.5,NaN,NaN,NaN,NaN,NaN,2.5,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
607,4.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
608,2.5,2.0,2.0,NaN,NaN,NaN,NaN,NaN,NaN,4.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [24]:
# Contar el número total de NaN en la matriz de escasez
total_nan = escasez_100K.isna().sum().sum()

print(f"Número total de valores NaN: {total_nan}")


Número total de valores NaN: 5830804


En total tenemos 5,830,804 valores NULOS en nuestra matriz, generando una abrumadora cantidad que tenemos que arreglar para poder generar nuestra algoritmo de CF